causalml/feature_selection/filters.py

```diff
@@ -0,0 +1,370 @@
+        if qk < 0.1**6:
+            qk = 0.1**6
+        elif qk > 1 - 0.1**6:
+            qk = 1 - 0.1**6
+        S = pk * np.log(pk / qk) + (1-pk) * np.log((1-pk) / (1-qk))
+        return S
+
+    def _evaluate_KL(self, nodeSummary, control_group='control'):
+        """
+        Calculate the multi-treatment unconditional D (one node)
+        with KL Divergence as split Evaluation function.
+
+        Parameters
+        ----------
+        control_group : str, optional (default='control')
+
+        Notes
+        -----
+        The function works for more than one non-control treatment groups.
+        """
+        if control_group not in nodeSummary:
+            return 0
+        pc = nodeSummary[control_group][0]
+        d_res = 0
+        for treatment_group in nodeSummary:
+            if treatment_group != control_group:
+                d_res += self._kl_divergence(nodeSummary[treatment_group][0], pc)
+        return d_res
+
+    def _evaluate_ED(self, nodeSummary, control_group='control'):
+        """
+        Calculate the multi-treatment unconditional D (one node)
+        with Euclidean Distance as split Evaluation function.
+        """
+        if control_group not in nodeSummary:
+            return 0
+        pc = nodeSummary[control_group][0]
+        d_res = 0
+        for treatment_group in nodeSummary:
+            if treatment_group != control_group:
+                d_res += 2 * (nodeSummary[treatment_group][0] - pc)**2
+        return d_res
+
+    def _evaluate_Chi(self, nodeSummary, control_group='control'):
+        """
+        Calculate the multi-treatment unconditional D (one node)
+        with Chi-Square as split Evaluation function.
+        """
+        if control_group not in nodeSummary:
+            return 0
+        pc = nodeSummary[control_group][0]
+        d_res = 0
+        for treatment_group in nodeSummary:
+            if treatment_group != control_group:
+                d_res += (
+                    (nodeSummary[treatment_group][0] - pc)**2 / max(0.1**6, pc) 
+                    + (nodeSummary[treatment_group][0] - pc)**2 / max(0.1**6, 1-pc)
+                )
+        return d_res
+
+
+    def _filter_D_one_feature(self, data, X_name, y_name, 
+                              n_bins=10, method='KL', control_group='control',
+                              experiment_group_column='treatment_group_key'):
+        """
+        Calculate the chosen divergence measure for one feature.
+        """
+        # [TODO] Application to categorical features
+
+        if method == 'KL':
+            evaluationFunction = self._evaluate_KL
+        elif method == 'ED':
+            evaluationFunction = self._evaluate_ED
+        elif method == 'Chi':
+            evaluationFunction = self._evaluate_Chi
+
+        totalSize = len(data.index)
+        x_bin = pd.qcut(data[X_name].values, n_bins, labels=False, 
+                        duplicates='drop')
```

Comment:
Do we want to silently drop the duplicates or eg add some jitter? Recall the problem with small sample sizes.

Reply:
Good call. One difficulty is that the scale of X is unknown, therefore, adding 10**(-8) may not be always safe. One option is to leave it to users to manipulate X outside, the other option, is to add a value that is proportional to the scale of X, e.g. at the scale of 10**(-8)*median(X).

Reply:
In the interest of explicitness, I think it may make sense to switch to duplicates='raise' and let the user manipulate X as needed.